[Woo POS][Product Search] Improve search UI

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of WOOMOB-147

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR tidies up the design of the search now that we have the new headers.

We don't actually have finalized designs yet, so this is generally based on what we've already done for coupons, what we've wireframed, and how things are on Android.

I noticed that we keep the view state of whether we're looking at products/coupons and whether we're searching or not between orders... this PR updates it so we also keep the search term at least, and I'll address the state reset in future work.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Enable the `searchProductsInPOS` flag and the `enableCouponsInPointOfSale` flags

1. Launch the app and navigate to POS
2. Tap the 🔍 button – observe that the titles are hidden, and the search field is focused.
3. Tap return on the keyboard – observe that you can use the pre-search state.
4. Search for a product and add it to the cart
5. Tap return on the keyboard – while still in search state, tap `Check out`
6. Tap back – observe that you're still in search
7. Tap back – observe that the search state is cleared and the titles shown again
8. Tap `Coupons` – observe that the search button is replaced with an `+` button

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I've tested on an iPad Air running iOS 17.7.3

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/6568c452-dc4a-4357-812c-fd998ce8a4ee


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
Use button to show or hide search field

joshheald committed:
Show and hide search field based on button taps

joshheald committed:
Hide headers when search is showing

joshheald committed:
Stay on pre-search state when return is tapped

joshheald committed:
Keep search state when coming back from the cart

If you go to the cart from a search state, then come back, the search term now stays in place.

Note that this isn’t only when coming back from the cart, it’s also when starting a new order, which isn’t great.

However, before this change it would keep the choice to show the search, but not the search term, so you’d just see the wrong items controller or the empty state.

I’ll improve it further in future work.

joshheald committed:
Always show the back button when searching

joshheald committed:
Animate changes between item list states

Note that we no longer change the selected list type based on the search term, it’s only by tapping on the search button or the back button. That code clearly wasn’t really doing anything now that the search field isn’t rendered when we go back… or something.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Limit header dynamic type sizes

The header now has potentially long text as the title/tab names. If allowed to grow too large, this will break the UI.

Since these elements are very consistent, it’s not such a problem to limit the size they can grow to.

joshheald committed:
Merge branch 'trunk' into woomob-147-improve-search-box-ui

joshheald commented:
How am I so bad at pressing the "request review" button? Sorry @staskus, this one actually comes before the one you just reviewed. No huge rush though.

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -114,48 +133,56 @@ private extension ItemListView {
-                                if !Task.isCancelled {
-                                    didFinishSearch = true
+                    if isSearchAllowed {
+                        searchField
```
Maybe it deserves to get into its own component? It has distinct logic from `ItemListView`. It could accept the searchTerm and call a callback when debounce is ready to make an API request.

joshheald replied:
Yeah, that's the plan. Just getting a little further before I factor it out 👍

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -114,48 +133,56 @@ private extension ItemListView {
+
+                        POSPageHeaderActionButton(systemName: "magnifyingglass") {
+                            selectedItemListType = .products(search: true)
+                            isSearchFieldFocused = true
```
The keyboard doesn't appear automatically the first time I tap the search button and there are no recent searches.

It appears automatically after I add the first recent search.

iPad Air M2 iOS 18.3

https://github.com/user-attachments/assets/5700dd67-6812-46ba-95c3-6968dc59891a

joshheald replied:
That's very strange... and I can't reproduce it.

Here's 18.4 on a simulator (you have to have used the software keyboard for the previous text entry in that case):

https://github.com/user-attachments/assets/fb7e9f4d-f120-4c09-826a-3489e38d2087

It's the same on iOS 18.2 simulator, and on 17.7.3 on my test iPad. I can't find an 18.3 simulator.

Is this 100%, or difficult to reproduce?

staskus replied:
It's quite hard to understand. Maybe it's some OS or SwiftUI bug? It's not 100% reproducible. At first, the keyboard opened, then I dismissed it, and after that, it wouldn't appear automatically anymore. I went to 'Menu -> Get Support' which triggered keyboard opening. After that, the keyboard continued to open even if I dismissed it.

https://github.com/user-attachments/assets/743507e5-353d-4659-8f87-ea377c36467c

I do have a physical Bluetooth keyboard for my iPad, but I tested it now with Bluetooth turned off. 

I think we can create a lower-priority ticket for later. I'll investigate it myself if you don't manage to reproduce it yourself.

joshheald replied:
@staskus Yeah, the hardware keyboard was my thought as well. Another potential culprit is Continuity, if you have it signed in to the same Apple Account as your Mac.

For me, the keyboard is shown immediately and consistently... but I'll hook up a keyboard and try a little more to recreate, and make an issue if I can't. 

We already specifically set the text field as focused when the button's tapped. There may be a timing issue with that on iOS 18, but it's a `@State` of the container view, so it should be respected regardless of the timing, really.

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -114,48 +133,56 @@ private extension ItemListView {
                         }
-                        .opacity(isAddingCouponAllowed ? 1 : 0)
+                        .renderedIf(isAddingCouponAllowed)
+                        .transition(.opacity.combined(with: .scale))
```
I used opacity to "reserve" a place for the button to avoid issues when switching between Products and Coupons with large dynamic type. Those are quite exceptional cases. There won't be issues while the number of buttons matches in Coupons and Products.

As I understand, some of that is solved by limiting the dynamic type size for the header?

staskus replied:
Either way, I love the animation!

joshheald replied:
Yep, I know, and agree that it's rare I looked at trying to keep the reserved space, but it was over complicated and didn't bring much value. 

It's mainly fixed by limiting the size of the titles.

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -114,48 +133,56 @@ private extension ItemListView {
-                                if !Task.isCancelled {
-                                    didFinishSearch = true
+                    if isSearchAllowed {
+                        searchField
```
Since we don't have designs, I'm not sure what will and what won't be addressed for this release.

However, I would at least expect a larger search field text.

joshheald replied:
Yeah good point, I'll make it a bit bigger. But for this release, it'll look pretty much as it does here. The main visual changes remaining are to add a clear button, update the empty state (in another PR already) and update the spacing of the grid items.

staskus reviewed (approved):
👍 Looks good.

Need to solve the keyboard issue before merging.

joshheald committed:
Ensure search field is on screen before focusing

joshheald committed:
Merge branch 'trunk' into woomob-147-improve-search-box-ui

# Conflicts:
#	WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift
#	WooCommerce/Classes/POS/Presentation/Reusable Views/POSPageHeaderView.swift
